arch/arm/core/userspace.S

```diff
@@ -73,6 +64,28 @@ SECTION_FUNC(TEXT,_arm_userspace_enter)
     msr PSPLIM, r0
 #endif
 
+#if defined (CONFIG_ARM_MPU)
+    /* Re-program dynamic memory map.
+     *
+     * Important note:
+     * _arch_configure_dynamic_mpu_regions() may re-program the MPU Stack Guard
+     * to guard the privilege stack for overflows (if building with option
+     * CONFIG_MPU_STACK_GUARD). There is a risk of actually overflowing the
+     * stack while doing the re-programming. We minimize the risk by placing
+     * this function immediately after we have switched to the privileged stack
+     * so that the whole stack area is available for this critical operation.
+     *
+     * Note that the risk for overflow is higher if using the normal thread
+     * stack, since we do not control how much stack is actually left, when
+     * user invokes _arm_userspace_enter().
+     */
+    push {r1,r2,r3,ip,lr}
```

Comment:
need 8 byte aligned pushes before function calls.  otherwise stack could go non 8 byte aligned.

Reply:
This is not exactly before the BL instruction though, @agross-linaro

Reply:
Fixed (Added r0, to make 8-byte aligned push)